src/frontend/src/optimizer/mod.rs

```diff
@@ -500,23 +500,75 @@ impl PlanRoot {
+    /// [`plan`] must be root node.
+    ///
+    /// NOTE: Optional `BatchTopN` / `BatchFilter`, are supported,
+    /// so less data sent over RPC.
+    fn enforce_exchange_above_table_scan(root: PlanRef, order: &Order) -> PlanRef {
+        fn is_candidate(plan: &PlanRef) -> bool {
+            // Unwrap optional `BatchTopN` / `BatchFilter`.
+            let mut candidate = plan.clone();
+            let mut inputs = candidate.inputs();
+            while candidate.node_type() == PlanNodeType::BatchTopN
+                || candidate.node_type() == PlanNodeType::BatchFilter
+            {
+                candidate = inputs[0].clone();
+                inputs = candidate.inputs();
+            }
+            // Table Scan
+            candidate.node_type() == PlanNodeType::BatchSource
+                || (candidate.node_type() == PlanNodeType::BatchSeqScan
+                    && !candidate
+                        .downcast_ref::<BatchSeqScan>()
+                        .unwrap()
+                        .logical()
+                        .is_sys_table())
+        }
+
+        fn is_exchange(plan: &PlanRef) -> bool {
+            plan.node_type() == PlanNodeType::BatchExchange
+        }
+
+        fn insert_exchange(plan: PlanRef, order: &Order) -> PlanRef {
+            BatchExchange::new(plan, order.clone(), Distribution::Single).into()
+        }
+
+        fn helper(plan: PlanRef, order: &Order) -> PlanRef {
+            if is_exchange(&plan) {
+                return plan;
+            }
+            if is_candidate(&plan) {
+                return insert_exchange(plan, order);
+            }
+            let new_inputs = plan
+                .inputs()
+                .into_iter()
+                .map(|input| helper(input, order))
+                .collect_vec();
+            plan.clone_with_inputs(&new_inputs)
+        }
+
+        if is_candidate(&root) {
+            insert_exchange(root, order)
+        } else {
+            helper(root, order)
+        }
+    }
+
     /// Optimize and generate a batch query plan for local execution.
     pub fn gen_batch_local_plan(&mut self) -> Result<PlanRef> {
         let mut plan = self.gen_batch_plan()?;
 
         // Convert to local plan node
         plan = plan.to_local_with_order_required(&self.required_order)?;
 
-        // We remark that since the `to_local_with_order_required` does not enforce single
-        // distribution, we enforce at the root if needed.
-        let insert_exchange = match plan.distribution() {
-            Distribution::Single => Self::require_additional_exchange_on_root(plan.clone()),
-            _ => true,
-        };
-        if insert_exchange {
-            plan =
-                BatchExchange::new(plan, self.required_order.clone(), Distribution::Single).into()
-        }
+        // Ensure there is exchange before all seq scan.
+        plan = Self::enforce_exchange_above_table_scan(plan, &self.required_order);
```

Comment:
IMHO, the enforcement here is trying to fix some unexpected plans at the end, while I think it could be more proper to handle it at the `to_local()` method of the `BatchSeqScan`. We can try to provide `SomeShard` for the `BatchSeqScan` in local mode, so that we can always enforce an exchange on the top of table scan and by the way we can utilize the optimization such as push filter/ project through the exchange.

Reply:
> IMHO, the enforcement here is trying to fix some unexpected plans at the end, while I think it could be more proper to handle it at the to_local() method of the BatchSeqScan and BatchSource. >

Good suggestion, I think that makes more sense.

> We can try to provide SomeShard for the BatchSeqScan in local mode, so that we can always enforce an exchange on the top of table scan

Hmm don't quite understand this. Is `SomeShard` required to enforce an exchange on top of table scan? Why is that so?

> by the way we can utilize the optimization such as push filter/ project through the exchange.

Yes I think this is good suggestion. Thanks.

Reply:
For local batch plans, almost all operators will require their input with `Singleton` distribution, while some `TableScan`'s distribution actually is `Singleton`. Our enforcement will think since the distribution is already satisfied, we can just skip placing an exchange operator here. But as we know, in local execution we need an exchange operator to keep table scan run in the CN. So I think we can hack the distribution for table scan when we call the `to_local()` method and return a new clone table scan with `SomeShard ` distribution so that the enforcement of the exchange will always work because we require a `Singleton` distribution while the table can can only provide the `SomeShard` distribution now.

Reply:
+1 for @chenzl25 's suggestion. The root cause of the issue in #7115 is that left table is a singleton, so we can't add exechange on top of table scan. I think in local mode we can always treats table scan as not single, so the overall process need not to change much.